add network addons operator

phoracek opened the pull request.

phoracek commented:
/hold

phoracek commented:
@rthallisey could you please take a look?

phoracek commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -136,6 +138,24 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
 		return result, err
 	}
 
+	// Define a new CDI object
```
Done

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -192,3 +212,21 @@ func newCDIForCR(cr *hcov1alpha1.HyperConverged) *cdi.CDI {
+	return &networkaddons.NetworkAddonsConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      networkaddonsnames.OPERATOR_CONFIG,
+			Namespace: networkaddonsnames.APPLIED_NAMESPACE,
```
The hyperconverged-cluster-operator is not (or at least should not) have permissions to be creating components across namespaces. Maybe I don't understand the direction that we are going but I would expect _User subscribes to the HCO and gets the HCO operator and the component operators (KubeVirt, CDI, ClusterNetworkAddons, etc.). Then a HCO CR would result in all of the components installed in the namespace where the CR is created._

djzager replied:
@rthallisey you may be most qualified to direct here.

phoracek replied:
Sorry this is a mistake, NetworkAddonsConfig is cluster-wide, non-namespaced resource. Will fix it.

There are two issues connected to your request:

1) Since network addons alter hosts, they can be deployed only once, therefore we have the resource cluster-wide and with a single acknowledged name "cluster". IIUIC, this is the same for CDI, therefore we could copy CDI leader election.
2) Network addons operator deploys third-party components (multus and linux bridge CNI) into their own namespaces since they are self-container and separated projects.

lveyde commented on README.md:
````diff
@@ -41,6 +41,14 @@ Launch the CDI operator.
+Launch the Cluster Network Addons operator.
+```bash
+oc create -f https://github.com/kubevirt/cluster-network-addons-operator/releases/download/v0.1.0/cluster-network-addons-operator_00_namespace.yaml
+oc create -f https://github.com/kubevirt/cluster-network-addons-operator/releases/download/v0.1.0/cluster-network-addons-operator_01_crd.yaml
````
Shouldn't the HCO Operator itself install these CRDs?

phoracek replied:
I think that it should eventually. However, now it has to be done manually. Correct me if I'm wrong please, I base this only on the readme.

rthallisey replied:
That's out of scope for the HCO.  OLM creates the CRDs.

phoracek commented:
I was not able to verify this PT as it is, since there seems to be a bug in CDI. However, after commenting out CDI parts from the operator, it deployed KubeVirt and network components OK.

phoracek commented:
/hold cancel

phoracek commented:
/hold

Hold this PR until we make it possible to run it namespaced.

phoracek commented:
/hold cancel

We cannot run in the same namespace as HCO. Network addons operator (and IIUIC CDI as well) needs to create a ClusterRoleBinding. The ClusterRoleBinding specifies ServiceAccount and its namespace. The namespace has to be hardcoded in operator's RBAC manifest. Therefore, until we are integrated with OLM which can adjust those manifests, we have no other choice but to deploy Network Addons operator in its own namespace.

@lveyde @djzager Please take a look and tell me whether there is something I should change in this PR.

phoracek commented:
@lveyde @rthallisey could you please take another look? I added network addons operator deployment to the hack/ directory and added support for imagePullPolicy

rthallisey commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -210,3 +245,21 @@ func newCDIForCR(cr *hcov1alpha1.HyperConverged) *cdi.CDI {
+		},
+		Spec: networkaddons.NetworkAddonsConfigSpec{
+			Multus:          &networkaddons.Multus{},
+			LinuxBridge:     &networkaddons.LinuxBridge{},
```
Why not expose Multus and LinuxBridge as configurable options too?

phoracek replied:
HCO should be opinionated. In my opinion Linux Bridge and Multus should be always deployed with this hyperconverged infrastructure.

rthallisey replied:
That's fine, but we can default these values behind the scenes.  I'm wondering if we should give the user the opportunity to configure them to be something other than the defaults.

phoracek replied:
Could we start with this and introduce configurable values in a different PR? I think it should be discussed with wider audience from the whole project (and it may also affect kubevirt and CDI). Where is the line between flexibility and "the right way to use hyperconverged cluster".

rthallisey replied:
@phoracek sounds good to me

rthallisey commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -154,6 +171,24 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+
+	// Check if this NetworkAddonsConfig CR already exists
+	foundNetworkAddons := &networkaddons.NetworkAddonsConfig{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: networkaddonsnames.OPERATOR_CONFIG, Namespace: ""}, foundNetworkAddons)
```
They are the same, but I'd prefer to use the `networkAddonsCR` to get the CR Name instead of grabbing it from the networkaddon code.  That way if we ever move away from the `OPERATOR_CONFIG` name, we'll have an abstraction that will pickup the change.

phoracek replied:
Done

phoracek commented on hack/operator-test.sh:
```diff
@@ -27,23 +27,26 @@ done
 set -x
 
-oc create -f ${TEMP_DIR}/
+./oc create -f ${TEMP_DIR}/
```
Oh right, forgot this from testing

phoracek replied:
Done

phoracek commented on hack/operator-test.sh:
```diff
@@ -27,23 +27,26 @@ done
 echo "Let the API server process the CRs"
 sleep 10
-oc wait kubevirt kubevirt --for condition=Ready -n kubevirt
+./oc wait kubevirt kubevirt --for condition=Ready -n kubevirt
```
I am working on status reporting for network addons. It's not ready yet though.

phoracek commented:
All comments should be addressed now.

@djzager can we merge this PR before #8 or do we need to wait for it?

phoracek commented:
Thank you! :)
